File: neoforge/src/main/java/com/stal111/forbidden_arcanus/data/enhancer/ModEnhancerDefinitions.java
package com.stal111.forbidden_arcanus.data.enhancer;

import com.stal111.forbidden_arcanus.ForbiddenArcanus;
import com.stal111.forbidden_arcanus.common.block.entity.forge.essence.EssenceType;
import com.stal111.forbidden_arcanus.common.item.enhancer.EnhancerDefinition;
import com.stal111.forbidden_arcanus.common.item.enhancer.EnhancerTarget;
import com.stal111.forbidden_arcanus.common.item.enhancer.condition.TimeCondition;
import com.stal111.forbidden_arcanus.common.item.enhancer.effect.MultiplyRequiredEssenceEffect;
import com.stal111.forbidden_arcanus.common.item.enhancer.effect.MultiplySoulDurationEffect;
import com.stal111.forbidden_arcanus.core.init.ModItems;
import com.stal111.forbidden_arcanus.core.registry.FARegistries;
import net.minecraft.data.worldgen.BootstrapContext;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceKey;
import net.valhelsia.valhelsia_core.api.common.registry.helper.datapack.DatapackRegistryClass;
import net.valhelsia.valhelsia_core.api.common.registry.helper.datapack.DatapackRegistryHelper;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * @author stal111
 * @since 2023-02-19
 */
public class ModEnhancerDefinitions extends DatapackRegistryClass<EnhancerDefinition> {

    public static final DatapackRegistryHelper<EnhancerDefinition> HELPER = ForbiddenArcanus.REGISTRY_MANAGER.getHelper(FARegistries.ENHANCER_DEFINITION);

    public static final ResourceKey<EnhancerDefinition> ARTISAN_RELIC = HELPER.createKey("artisan_relic");
    public static final ResourceKey<EnhancerDefinition> CRESCENT_MOON = HELPER.createKey("crescent_moon");
    public static final ResourceKey<EnhancerDefinition> CRIMSON_STONE = HELPER.createKey("crimson_stone");
    public static final ResourceKey<EnhancerDefinition> SOUL_CRIMSON_STONE = HELPER.createKey("soul_crimson_stone");
    public static final ResourceKey<EnhancerDefinition> ELEMENTARIUM = HELPER.createKey("elementarium");
    public static final ResourceKey<EnhancerDefinition> DIVINE_PACT = HELPER.createKey("divine_pact");
    public static final ResourceKey<EnhancerDefinition> MALEDICTUS_PACT = HELPER.createKey("maledictus_pact");

    public ModEnhancerDefinitions(BootstrapContext<EnhancerDefinition> context) {
        super(context);
    }

    @Override
    public void bootstrap(BootstrapContext<EnhancerDefinition> context) {
        context.register(ARTISAN_RELIC, EnhancerDefinition.create(this.generateDescription("artisan_relic", EnhancerTarget.HEPHAESTUS_FORGE, EnhancerTarget.CLIBANO), ModItems.ARTISAN_RELIC.get(), new MultiplyRequiredEssenceEffect(List.of(), EssenceType.EXPERIENCE, 0.75)));
        context.register(CRESCENT_MOON, EnhancerDefinition.create(this.generateDescription("crescent_moon", EnhancerTarget.HEPHAESTUS_FORGE), ModItems.CRESCENT_MOON.get(), new MultiplyRequiredEssenceEffect(List.of(new TimeCondition(23000, 23999)), EssenceType.AUREAL, 0.9), new MultiplyRequiredEssenceEffect(List.of(new TimeCondition(0, 12999)), EssenceType.AUREAL, 0.9), new MultiplyRequiredEssenceEffect(List.of(new TimeCondition(13000, 22999)), EssenceType.AUREAL, 0.99)));
        context.register(CRIMSON_STONE, EnhancerDefinition.create(this.generateDescription("crimson_stone", EnhancerTarget.HEPHAESTUS_FORGE, EnhancerTarget.CLIBANO), ModItems.CRIMSON_STONE.get(), new MultiplyRequiredEssenceEffect(List.of(), EssenceType.SOULS, 0.5), new MultiplySoulDurationEffect(List.of(), 1.3D)));
        context.register(SOUL_CRIMSON_STONE, EnhancerDefinition.create(this.generateDescription("soul_crimson_stone", EnhancerTarget.HEPHAESTUS_FORGE), ModItems.SOUL_CRIMSON_STONE.get(), new MultiplyRequiredEssenceEffect(List.of(), EssenceType.AUREAL, 0.0), new MultiplyRequiredEssenceEffect(List.of(), EssenceType.SOULS, 0.0), new MultiplyRequiredEssenceEffect(List.of(), EssenceType.BLOOD, 0.0), new MultiplyRequiredEssenceEffect(List.of(), EssenceType.EXPERIENCE, 0.0)));
        context.register(ELEMENTARIUM, EnhancerDefinition.create(this.generateDescription("elementarium", EnhancerTarget.HEPHAESTUS_FORGE), ModItems.ELEMENTARIUM.get()));
        context.register(DIVINE_PACT, EnhancerDefinition.create(this.generateDescription("divine_pact", EnhancerTarget.HEPHAESTUS_FORGE), ModItems.DIVINE_PACT.get()));
        context.register(MALEDICTUS_PACT, EnhancerDefinition.create(this.generateDescription("maledictus_pact", EnhancerTarget.HEPHAESTUS_FORGE), ModItems.MALEDICTUS_PACT.get()));
    }

    private Map<EnhancerTarget, Component> generateDescription(String key, EnhancerTarget... targets) {
        return Arrays.stream(targets).collect(Collectors.toMap(Function.identity(), enhancerTarget -> Component.translatable("item.forbidden_arcanus.enhancer." + key + "." + enhancerTarget.getSerializedName())));
    }
}

File: datagen/src/main/kotlin/com/stal111/forbidden_arcanus/datagen/item/ModEnhancerDefinitions.kt
package com.stal111.forbidden_arcanus.datagen.item

import com.stal111.forbidden_arcanus.common.essence.EssenceType
import com.stal111.forbidden_arcanus.common.item.enhancer.BuiltInEnhancers
import com.stal111.forbidden_arcanus.common.item.enhancer.EnhancerDefinition
import com.stal111.forbidden_arcanus.common.item.enhancer.EnhancerTarget
import com.stal111.forbidden_arcanus.common.item.enhancer.condition.EffectCondition
import com.stal111.forbidden_arcanus.common.item.enhancer.condition.TimeCondition
import com.stal111.forbidden_arcanus.common.item.enhancer.effect.ConditionalEnhancerEffect
import com.stal111.forbidden_arcanus.common.item.enhancer.effect.EnhancerEffect
import com.stal111.forbidden_arcanus.common.item.enhancer.effect.MultiplyRequiredEssenceEffect
import com.stal111.forbidden_arcanus.common.item.enhancer.effect.MultiplySoulDurationEffect
import com.stal111.forbidden_arcanus.core.init.ModItems
import net.minecraft.data.worldgen.BootstrapContext
import net.minecraft.network.chat.Component
import net.valhelsia.dataforge.RegistryDataProvider
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors

object ModEnhancerDefinitions : RegistryDataProvider<EnhancerDefinition> {
    override fun bootstrap(context: BootstrapContext<EnhancerDefinition>) {
        context.register(
            BuiltInEnhancers.ARTISAN_RELIC,
            EnhancerDefinition.create(
                generateDescription("artisan_relic", EnhancerTarget.HEPHAESTUS_FORGE, EnhancerTarget.CLIBANO),
                ModItems.ARTISAN_RELIC.get(),
                effect(MultiplyRequiredEssenceEffect(EssenceType.EXPERIENCE, 0.75))
            )
        )
        context.register(
            BuiltInEnhancers.CRESCENT_MOON,
            EnhancerDefinition.create(
                generateDescription("crescent_moon", EnhancerTarget.HEPHAESTUS_FORGE),
                ModItems.CRESCENT_MOON.get(),
                effect(
                    MultiplyRequiredEssenceEffect(EssenceType.AUREAL, 0.9),
                    TimeCondition(23000, 23999)
                ),
                effect(
                    MultiplyRequiredEssenceEffect(EssenceType.AUREAL, 0.9),
                    TimeCondition(0, 12999)
                ),
                effect(
                    MultiplyRequiredEssenceEffect(EssenceType.AUREAL, 0.99),
                    TimeCondition(13000, 22999)
                )
            )
        )
        context.register(
            BuiltInEnhancers.CRIMSON_STONE,
            EnhancerDefinition.create(
                generateDescription("crimson_stone", EnhancerTarget.HEPHAESTUS_FORGE, EnhancerTarget.CLIBANO),
                ModItems.CRIMSON_STONE.get(),
                effect(MultiplyRequiredEssenceEffect(EssenceType.ECTOPLASM, 0.5)),
                effect(MultiplySoulDurationEffect(1.3))
            )
        )
        context.register(
            BuiltInEnhancers.SOUL_CRIMSON_STONE,
            EnhancerDefinition.create(
                generateDescription("soul_crimson_stone", EnhancerTarget.HEPHAESTUS_FORGE),
                ModItems.SOUL_CRIMSON_STONE.get(),
                effect(MultiplyRequiredEssenceEffect(EssenceType.AUREAL, 0.0)),
                effect(MultiplyRequiredEssenceEffect(EssenceType.ECTOPLASM, 0.0)),
                effect(MultiplyRequiredEssenceEffect(EssenceType.BLOOD, 0.0)),
                effect(MultiplyRequiredEssenceEffect(EssenceType.EXPERIENCE, 0.0))
            )
        )
        context.register(
            BuiltInEnhancers.ELEMENTARIUM,
            EnhancerDefinition.create(
                generateDescription("elementarium", EnhancerTarget.HEPHAESTUS_FORGE),
                ModItems.ELEMENTARIUM.get()
            )
        )
        context.register(
            BuiltInEnhancers.DIVINE_PACT,
            EnhancerDefinition.create(
                generateDescription("divine_pact", EnhancerTarget.HEPHAESTUS_FORGE),
                ModItems.DIVINE_PACT.get()
            )
        )
        context.register(
            BuiltInEnhancers.MALEDICTUS_PACT,
            EnhancerDefinition.create(
                generateDescription("maledictus_pact", EnhancerTarget.HEPHAESTUS_FORGE),
                ModItems.MALEDICTUS_PACT.get()
            )
        )
    }

    private fun generateDescription(
        key: String,
        vararg targets: EnhancerTarget
    ): MutableMap<EnhancerTarget, Component> {
        return Arrays.stream(targets).collect(
            Collectors.toMap(Function.identity()) { Component.translatable("item.forbidden_arcanus.enhancer." + key + "." + it.serializedName) }
        )
    }

    private fun <T : EnhancerEffect> effect(effect: T): ConditionalEnhancerEffect<T> {
        return ConditionalEnhancerEffect.of<T>(effect)
    }

    private fun <T : EnhancerEffect> effect(effect: T, condition: EffectCondition): ConditionalEnhancerEffect<T> {
        return ConditionalEnhancerEffect.of<T>(effect, listOf(condition))
    }
}
